lib/core/util.c

```diff
@@ -945,3 +946,60 @@ void h2o_cleanup_thread(void)
     h2o_mem_clear_recycle(&h2o_http2_wbuf_buffer_prototype.allocator);
     h2o_mem_clear_recycle(&h2o_socket_buffer_prototype.allocator);
 }
+
+void h2o_generate_uuidv4(char *buf)
+{
+    // RFC-4122 4.4. Algorithms for Creating a UUID from Truly Random or Pseudo-Random Numbers
+
+    // 4.2.1 Layout And Byte Order
+    struct rfc4122 {
+        uint32_t time_low;
+        uint16_t time_mid;
+        uint16_t time_hi_and_version;
+        uint8_t clock_seq_hi_and_reserved;
+        uint8_t clock_seq_low;
+        uint8_t node[6];
+    } uuid;
+    _Static_assert(sizeof(uuid) == 16, "size of UUID structure must be 16");
+
+    if (RAND_bytes((void*)&uuid, sizeof(uuid)) != 1) {
+        h2o_fatal("RAND_bytes failed");
+    }
+
+    // Variant:
+    // > Set the two most significant bits (bits 6 and 7) of the
+    // > clock_seq_hi_and_reserved to zero and one, respectively.
+    uuid.clock_seq_hi_and_reserved = (uuid.clock_seq_hi_and_reserved & 0x3f) & 0x80;
+
+    // Version:
+    // > Set the four most significant bits (bits 12 through 15) of the
+    // > time_hi_and_version field to the 4-bit version number from
+    // > Section 4.1.3.
+    const int version = 4;
+    uuid.time_hi_and_version = (uuid.time_hi_and_version & 0x0f) | (version << 4);
```

Comment:
Is this correct?

I believe that the value is expected to preserve 12 bits of random with remaining 4 bits used to represent the version number. But my reading of this statement is that only 4 bits of random is being preserved.

I appreciate your efforts to follow what the specification says, but I wonder if it would actually be better to simply do everything flattened all the time; i.e. generate random bytes in a byte array (`uint8_t [16]`), flip the necessary bits, then serialize it.

Reply:
I thought using struct was much simpler than `uint8_t[16]`, but it has turned out you are right -- I was confused by the byte order.